File: backend/app/routers/events.py
"""Event creation, discovery, detail, RSVP, attendees, and co-host endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.core.config import Settings, get_settings
from app.core.db import acquire
from app.core.errors import api_error
from app.core.security import get_current_user_id
from app.repositories import events_repo, users_repo
from app.schemas.events import (
    Attendee,
    AttendeeList,
    CreateEventRequest,
    EventDetail,
    EventList,
    RsvpRequest,
    RsvpResponse,
    UpdateEventRequest,
)
from app.services.location_service import clamp_radius
from app.services.pagination import decode_cursor, encode_cursor

router = APIRouter(prefix="/events", tags=["events"])

_HOST_ROLES = {"host", "co-host"}


def _event_detail(row, *, distance_m=None, my_rsvp=None) -> EventDetail:
    keys = row.keys()
    if distance_m is None and "distance_m" in keys and row["distance_m"] is not None:
        distance_m = row["distance_m"]
    if my_rsvp is None and "my_rsvp" in keys:
        my_rsvp = row["my_rsvp"]
    return EventDetail(
        id=str(row["id"]),
        creator_id=str(row["creator_id"]),
        title=row["title"],
        description=row["description"],
        lat=row["lat"],
        lng=row["lng"],
        radius_m=row["radius_m"],
        capacity=row["capacity"],
        tags=list(row["tags"] or []),
        visibility=row["visibility"],
        starts_at=row["starts_at"],
        ends_at=row["ends_at"],
        is_archived=row["is_archived"],
        attendee_count=row["attendee_count"] if "attendee_count" in keys else 0,
        distance_m=round(distance_m, 1) if distance_m is not None else None,
        my_rsvp=my_rsvp,
    )


@router.post("", response_model=EventDetail, status_code=201)
async def create_event(
    body: CreateEventRequest,
    user_id: str = Depends(get_current_user_id),
) -> EventDetail:
    async with acquire(user_id) as conn:
        row = await events_repo.create_event(
            conn, user_id,
            title=body.title, description=body.description,
            lat=body.lat, lng=body.lng, radius_m=body.radius_m,
            capacity=body.capacity, tags=body.tags, visibility=body.visibility,
            starts_at=body.starts_at, ends_at=body.ends_at,
        )
    detail = _event_detail(row, my_rsvp="going")
    detail.attendee_count = 1
    return detail


@router.get("", response_model=EventList)
async def list_events(
    user_id: str = Depends(get_current_user_id),
    lat: float = Query(..., ge=-90, le=90),
    lng: float = Query(..., ge=-180, le=180),
    radius: float | None = Query(default=None),
    tags: list[str] | None = Query(default=None),
    cursor: str | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=50),
    settings: Settings = Depends(get_settings),
) -> EventList:
    radius_m = clamp_radius(
        radius, min_m=settings.radius_min_m, max_m=settings.radius_max_m,
        default_m=settings.radius_default_m,
    )
    cur = decode_cursor(cursor)
    after_distance = cur.get("distance_m") if cur else None
    after_id = cur.get("id") if cur else None

    async with acquire(user_id) as conn:
        rows = await events_repo.list_nearby(
            conn, user_id, lat=lat, lng=lng, radius_m=radius_m, tags=tags,
            after_distance=after_distance, after_id=after_id, limit=limit,
        )

    items = [_event_detail(r) for r in rows]
    next_cursor = (
        encode_cursor({"distance_m": rows[-1]["distance_m"], "id": str(rows[-1]["id"])})
        if len(rows) == limit
        else None
    )
    return EventList(items=items, next_cursor=next_cursor)


@router.get("/{event_id}", response_model=EventDetail)
async def get_event(event_id: str, user_id: str = Depends(get_current_user_id)) -> EventDetail:
    async with acquire(user_id) as conn:
        row = await events_repo.get_event(conn, event_id, user_id)
    if row is None:
        raise api_error(404, "not_found", "Event not found")
    return _event_detail(row)


@router.put("/{event_id}", response_model=EventDetail)
async def update_event(
    event_id: str,
    body: UpdateEventRequest,
    user_id: str = Depends(get_current_user_id),
) -> EventDetail:
    async with acquire(user_id) as conn:
        role = await events_repo.role_for(conn, event_id, user_id)
        if role not in _HOST_ROLES:
            raise api_error(403, "forbidden", "Host only")
        fields = body.model_dump(exclude_none=True)
        row = await events_repo.update_event(conn, event_id, fields)
        if row is None:
            raise api_error(404, "not_found", "Event not found")
        full = await events_repo.get_event(conn, event_id, user_id)
    return _event_detail(full)


@router.post("/{event_id}/rsvp", response_model=RsvpResponse)
async def rsvp(
    event_id: str,
    body: RsvpRequest,
    user_id: str = Depends(get_current_user_id),
) -> RsvpResponse:
    async with acquire(user_id) as conn:
        if body.status == "going":
            capacity = await events_repo.event_capacity(conn, event_id)
            if capacity is not None:
                others = await events_repo.going_count_excluding(conn, event_id, user_id)
                if others >= capacity:
                    raise api_error(409, "at_capacity", "Event is full")
        row = await events_repo.upsert_rsvp(conn, event_id, user_id, body.status)
    return RsvpResponse(event_id=str(row["event_id"]), status=row["status"], role=row["role"])


@router.get("/{event_id}/attendees", response_model=AttendeeList)
async def attendees(event_id: str, user_id: str = Depends(get_current_user_id)) -> AttendeeList:
    """Full attendee list only for host/co-host; others get the count only (Requirement 13.4)."""
    async with acquire(user_id) as conn:
        role = await events_repo.role_for(conn, event_id, user_id)
        count = await events_repo.attendee_count(conn, event_id)
        if role in _HOST_ROLES:
            rows = await events_repo.list_attendees(conn, event_id)
            attendee_list = [
                Attendee(
                    user_id=str(r["user_id"]), display_name=r["display_name"],
                    avatar_url=r["avatar_url"], role=r["role"], status=r["status"],
                )
                for r in rows
            ]
            return AttendeeList(event_id=event_id, attendee_count=count, attendees=attendee_list)
    return AttendeeList(event_id=event_id, attendee_count=count, attendees=None)


@router.post("/{event_id}/cohosts/{cohost_id}", status_code=204)
async def add_cohost(
    event_id: str,
    cohost_id: str,
    user_id: str = Depends(get_current_user_id),
) -> None:
    """Promote a user to co-host (host only). Accepts the deep-link invite flow's target."""
    async with acquire(user_id) as conn:
        role = await events_repo.role_for(conn, event_id, user_id)
        if role != "host":
            raise api_error(403, "forbidden", "Host only")
        await users_repo.get_public_profile(conn, cohost_id)  # 404 surfaces via FK if absent
        await events_repo.set_role(conn, event_id, cohost_id, "co-host")

File: functions/src/events.ts
import { onCall } from "firebase-functions/v2/https";

import { MAX_EVENT_TAGS, clampRadius } from "./config";
import {
  db,
  usersCol,
  eventsCol,
  rsvpsCol,
  Timestamp,
} from "./lib/firestore";
import {
  requireUid,
  invalidArg,
  notFound,
  permissionDenied,
  failedPrecondition,
} from "./lib/errors";
import { geohashFor, haversineM, queryBounds, LatLng } from "./lib/geo";
import { RsvpStatus } from "./lib/types";

const HOST_ROLES = new Set(["host", "co-host"]);
const VALID_STATUS = new Set(["going", "maybe", "no"]);

function eventDetail(
  id: string,
  d: FirebaseFirestore.DocumentData,
  extra: { distanceM?: number; myRsvp?: string | null } = {},
) {
  return {
    id,
    creatorId: d.creatorId,
    title: d.title,
    description: d.description ?? null,
    lat: d.lat,
    lng: d.lng,
    radiusM: d.radiusM,
    capacity: d.capacity ?? null,
    tags: d.tags ?? [],
    visibility: d.visibility,
    startsAt: d.startsAt.toDate().toISOString(),
    endsAt: d.endsAt.toDate().toISOString(),
    isArchived: d.isArchived ?? false,
    attendeeCount: d.attendeeCount ?? 0,
    distanceM: extra.distanceM ?? null,
    myRsvp: extra.myRsvp ?? null,
  };
}

/** Create a geofenced event. Enforces ≤5 tags and starts < ends. */
export const createEvent = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const d = req.data ?? {};

  const title = String(d.title ?? "").trim();
  if (!title) throw invalidArg("title is required");
  const tags: string[] = Array.isArray(d.tags) ? d.tags : [];
  if (tags.length > MAX_EVENT_TAGS) throw invalidArg(`At most ${MAX_EVENT_TAGS} tags`);
  if (typeof d.lat !== "number" || typeof d.lng !== "number") {
    throw invalidArg("lat and lng are required");
  }
  const startsAt = new Date(d.startsAt);
  const endsAt = new Date(d.endsAt);
  if (Number.isNaN(startsAt.getTime()) || Number.isNaN(endsAt.getTime())) {
    throw invalidArg("startsAt and endsAt must be valid dates");
  }
  if (startsAt >= endsAt) throw invalidArg("startsAt must be before endsAt");

  const ref = eventsCol().doc();
  await ref.set({
    creatorId: uid,
    title,
    description: d.description ? String(d.description).slice(0, 2000) : null,
    lat: d.lat,
    lng: d.lng,
    geohash: geohashFor(d.lat, d.lng),
    radiusM: typeof d.radiusM === "number" ? d.radiusM : 500,
    capacity: typeof d.capacity === "number" ? d.capacity : null,
    tags,
    visibility: d.visibility === "invite-only" ? "invite-only" : "public",
    startsAt: Timestamp.fromDate(startsAt),
    endsAt: Timestamp.fromDate(endsAt),
    isArchived: false,
    attendeeCount: 1,
    createdAt: Timestamp.now(),
  });
  // Creator becomes the host (going).
  await rsvpsCol(ref.id).doc(uid).set({
    role: "host",
    status: "going",
    createdAt: Timestamp.now(),
  });

  const snap = await ref.get();
  return eventDetail(ref.id, snap.data()!, { myRsvp: "going" });
});

/** Nearby active events, distance-sorted, cursor-paginated. */
export const listEvents = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const d = req.data ?? {};
  if (typeof d.lat !== "number" || typeof d.lng !== "number") {
    throw invalidArg("lat and lng are required");
  }
  const center: LatLng = { lat: d.lat, lng: d.lng };
  const radiusM = clampRadius(typeof d.radius === "number" ? d.radius : undefined);
  const filterTags: string[] = Array.isArray(d.tags) ? d.tags : [];
  const now = Date.now();

  const bounds = queryBounds(center, radiusM);
  const snaps = await Promise.all(
    bounds.map((b) =>
      eventsCol().orderBy("geohash").startAt(b[0]).endAt(b[1]).get(),
    ),
  );

  type Row = { id: string; data: FirebaseFirestore.DocumentData; distanceM: number };
  const rows: Row[] = [];
  const seen = new Set<string>();
  for (const snap of snaps) {
    for (const doc of snap.docs) {
      if (seen.has(doc.id)) continue;
      seen.add(doc.id);
      const e = doc.data();
      if (e.isArchived) continue;
      if (e.endsAt && e.endsAt.toMillis() < now) continue;
      if (e.visibility !== "public" && e.creatorId !== uid) continue;
      const distanceM = haversineM(center, { lat: e.lat, lng: e.lng });
      if (distanceM > radiusM) continue;
      if (filterTags.length > 0) {
        const tagSet = new Set<string>(e.tags ?? []);
        if (!filterTags.some((t) => tagSet.has(t))) continue;
      }
      rows.push({ id: doc.id, data: e, distanceM });
    }
  }

  rows.sort((a, b) =>
    a.distanceM !== b.distanceM ? a.distanceM - b.distanceM : a.id < b.id ? -1 : 1,
  );

  const limit = Math.min(Math.max(Number(d.limit) || 20, 1), 50);
  const cursor = d.cursor as { distanceM: number; id: string } | undefined;
  let filtered = rows;
  if (cursor) {
    filtered = rows.filter(
      (r) =>
        r.distanceM > cursor.distanceM ||
        (r.distanceM === cursor.distanceM && r.id > cursor.id),
    );
  }
  const page = filtered.slice(0, limit);

  // Attach myRsvp for the page.
  const myRsvps = await Promise.all(
    page.map((r) => rsvpsCol(r.id).doc(uid).get()),
  );
  const items = page.map((r, i) =>
    eventDetail(r.id, r.data, {
      distanceM: Math.round(r.distanceM * 10) / 10,
      myRsvp: myRsvps[i].exists ? (myRsvps[i].get("status") as string) : null,
    }),
  );

  const last = page[page.length - 1];
  const nextCursor =
    page.length === limit && filtered.length > limit
      ? { distanceM: last.distanceM, id: last.id }
      : null;

  return { items, nextCursor };
});

/** Single event detail. */
export const getEvent = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const eventId = (req.data ?? {}).eventId as string;
  if (!eventId) throw invalidArg("eventId is required");
  const [snap, rsvp] = await Promise.all([
    eventsCol().doc(eventId).get(),
    rsvpsCol(eventId).doc(uid).get(),
  ]);
  if (!snap.exists) throw notFound("Event not found");
  return eventDetail(eventId, snap.data()!, {
    myRsvp: rsvp.exists ? (rsvp.get("status") as string) : null,
  });
});

/** RSVP (going/maybe/no). Maintains attendeeCount + capacity in a transaction. */
export const rsvpEvent = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const { eventId, status } = req.data ?? {};
  if (!eventId) throw invalidArg("eventId is required");
  if (!VALID_STATUS.has(status)) throw invalidArg("status must be going|maybe|no");

  const result = await db.runTransaction(async (tx) => {
    const evRef = eventsCol().doc(eventId);
    const ev = await tx.get(evRef);
    if (!ev.exists) throw notFound("Event not found");

    const rsvpRef = rsvpsCol(eventId).doc(uid);
    const prev = await tx.get(rsvpRef);
    const prevStatus = prev.exists ? (prev.get("status") as string) : null;
    const role = prev.exists ? (prev.get("role") as string) : "guest";

    const capacity = ev.get("capacity") as number | null;
    const count = (ev.get("attendeeCount") as number) ?? 0;
    const wasGoing = prevStatus === "going";
    const willGo = status === "going";

    if (willGo && !wasGoing && capacity != null && count >= capacity) {
      throw failedPrecondition("Event is full");
    }

    tx.set(rsvpRef, {
      role,
      status,
      createdAt: prev.exists ? prev.get("createdAt") : Timestamp.now(),
    });

    let delta = 0;
    if (willGo && !wasGoing) delta = 1;
    else if (!willGo && wasGoing) delta = -1;
    if (delta !== 0) tx.update(evRef, { attendeeCount: count + delta });

    return { role, status };
  });

  return { eventId, status: result.status as RsvpStatus, role: result.role };
});

/** Attendee list — full list for host/co-host only; otherwise just the count. */
export const eventAttendees = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const eventId = (req.data ?? {}).eventId as string;
  if (!eventId) throw invalidArg("eventId is required");

  const myRsvp = await rsvpsCol(eventId).doc(uid).get();
  const role = myRsvp.exists ? (myRsvp.get("role") as string) : null;
  const all = await rsvpsCol(eventId).get();
  const goingCount = all.docs.filter((r) => r.get("status") === "going").length;

  if (role && HOST_ROLES.has(role)) {
    const attendees = await Promise.all(
      all.docs.map(async (r) => {
        const u = await usersCol().doc(r.id).get();
        return {
          userId: r.id,
          displayName: u.get("displayName") ?? null,
          avatarUrl: u.get("avatarUrl") ?? null,
          role: r.get("role"),
          status: r.get("status"),
        };
      }),
    );
    return { eventId, attendeeCount: goingCount, attendees };
  }
  return { eventId, attendeeCount: goingCount, attendees: null };
});

/** Promote a user to co-host (host only). */
export const addCohost = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const { eventId, userId } = req.data ?? {};
  if (!eventId || !userId) throw invalidArg("eventId and userId are required");

  const mine = await rsvpsCol(eventId).doc(uid).get();
  if (!mine.exists || mine.get("role") !== "host") {
    throw permissionDenied("Host only");
  }
  await rsvpsCol(eventId).doc(userId).set(
    { role: "co-host", status: "going", createdAt: Timestamp.now() },
    { merge: true },
  );
  return { eventId, cohostId: userId };
});

/** Update an event (host/co-host only). */
export const updateEvent = onCall(async (req) => {
  const uid = requireUid(req.auth);
  const d = req.data ?? {};
  const eventId = d.eventId as string;
  if (!eventId) throw invalidArg("eventId is required");

  const evSnap = await eventsCol().doc(eventId).get();
  if (!evSnap.exists) throw notFound("Event not found");
  const myRsvp = await rsvpsCol(eventId).doc(uid).get();
  const role = myRsvp.exists ? (myRsvp.get("role") as string) : null;
  if (evSnap.get("creatorId") !== uid && !(role && HOST_ROLES.has(role))) {
    throw permissionDenied("Host only");
  }

  const update: Record<string, unknown> = {};
  if (d.title !== undefined) update.title = String(d.title).trim();
  if (d.description !== undefined) update.description = String(d.description).slice(0, 2000);
  if (d.capacity !== undefined) update.capacity = d.capacity;
  if (d.visibility !== undefined) {
    update.visibility = d.visibility === "invite-only" ? "invite-only" : "public";
  }
  if (d.tags !== undefined) {
    if (!Array.isArray(d.tags) || d.tags.length > MAX_EVENT_TAGS) {
      throw invalidArg(`At most ${MAX_EVENT_TAGS} tags`);
    }
    update.tags = d.tags;
  }
  let startsAt = evSnap.get("startsAt") as FirebaseFirestore.Timestamp;
  let endsAt = evSnap.get("endsAt") as FirebaseFirestore.Timestamp;
  if (d.startsAt !== undefined) {
    startsAt = Timestamp.fromDate(new Date(d.startsAt));
    update.startsAt = startsAt;
  }
  if (d.endsAt !== undefined) {
    endsAt = Timestamp.fromDate(new Date(d.endsAt));
    update.endsAt = endsAt;
  }
  if (startsAt.toMillis() >= endsAt.toMillis()) {
    throw invalidArg("startsAt must be before endsAt");
  }

  await eventsCol().doc(eventId).set(update, { merge: true });
  const snap = await eventsCol().doc(eventId).get();
  return eventDetail(eventId, snap.data()!, {
    myRsvp: myRsvp.exists ? (myRsvp.get("status") as string) : null,
  });
});
